aptos-move/move-examples/fungible_asset/sources/fungible_asset.move

```diff
@@ -0,0 +1,377 @@
+        /// The balance of the fungible asset.
+        balance: u64,
+        /// Whether this account is frozen for this fungible asset.
+        frozen: bool,
+        /// The delete_ref of this object, used for cleanup.
+        delete_ref: DeleteRef
+    }
+
+    /// The unpinned version of fungible asset.
+    /// Note: it does not have `store` ability so only used in hot potato pattern.
+    struct FungibleAsset {
+        asset_addr: address,
+        balance: u64,
+    }
+
+    /// Check the amount of an account.
+    public fun balance<T: key>(
+        fungible_asset_owner: address,
+        fungible_source: &Object<T>
+    ): u64 acquires FungibleAssetStore, PinnedFungibleAsset {
+        let pfa_opt = get_pinned_fungible_asset_object(
+            fungible_asset_owner,
+            object_address(fungible_source),
+            false
+        );
+        if (option::is_none(&pfa_opt)) {
+            return 0
+        };
+        let pfa = option::destroy_some(pfa_opt);
+        borrow_global<PinnedFungibleAsset>(object_address(&pfa)).balance
+    }
+
+    /// Check the coin account of `fungible_asset_owner` is frozen or not.
+    public fun is_frozen<T: key>(
+        fungible_asset_owner: address,
+        fungible_source: &Object<T>
+    ): bool acquires FungibleAssetStore, PinnedFungibleAsset {
+        let pfa_opt = get_pinned_fungible_asset_object(
+            fungible_asset_owner,
+            object_address(fungible_source),
+            false
+        );
+        if (option::is_none(&pfa_opt)) {
+            return false
+        };
+        let pfa = option::destroy_some(pfa_opt);
+        borrow_global<PinnedFungibleAsset>(object_address(&pfa)).frozen
+    }
+
+    /// Deposit fungible asset to an account.
+    public fun deposit(
+        fa: FungibleAsset,
+        to: address
+    ) acquires FungibleAssetStore, PinnedFungibleAsset {
+        let pfa = borrow_fungible_asset_mut(
+            to,
+            fa.asset_addr,
+            true /* create token object if not exist */
+        );
+        assert!(!pfa.frozen, error::invalid_argument(ESTILL_FROZEN));
+        merge(pfa, fa);
+    }
+
+    /// Mint fungible asset with `amount`.
+    public(friend) fun mint(
+        asset_addr: address,
+        amount: u64,
+    ): FungibleAsset {
+        assert!(amount != 0, error::invalid_argument(EAMOUNT_CANNOT_BE_ZERO));
+        FungibleAsset {
+            asset_addr,
+            balance: amount
+        }
+    }
+
+    /// Freeeze/unfreeze any account of asset address `asset_addr`.
+    public(friend) fun set_frozen_flag(
+        fungible_asset_owner: address,
+        asset_addr: address,
+        frozen: bool
+    ) acquires FungibleAssetStore, PinnedFungibleAsset {
+        let pfa_opt = get_pinned_fungible_asset_object(fungible_asset_owner, asset_addr, true);
+        let pfa = option::destroy_some(pfa_opt);
+        borrow_global_mut<PinnedFungibleAsset>(object_address(&pfa)).frozen = frozen;
+    }
```

Comment:
we're going down a dangerous path here, we should push the concept of the TransferRef into here rather than re-invent.. then fungible assets could determine what it means to freeze, enable allowlists, partial freeze, approvals, etc.

Reply:
I understand what you mean here. But it is really mind-blowing that how to offer the flexibility of all the features you mentioned here. lemme think about it.

Reply:
Don't overthink it, we already have the TransferRef pattern and ungated transfers, just replicate it :)

Reply:
only for partial freeze. I am confused that how transferRef pattern deal with this case.